Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact
the validity of the results.

1.Normality–Each sample was drawn from a normally distributed population.
2.Equal Variances – The variances of the populations that the samples come from are equal.
3.Independence – The observations in each group are independent of each other and the observations within groups were obtained by a random sample.

If these assumptions aren’t met, then the results of our one-way ANOVA could be unreliable.

Q2. What are the three types of ANOVA, and in what situations would each be used?



One-way ANOVA- One-way ANOVA is used to determine one factor with atleast 2 levels and these levels are independent.
Repeated measure ANOVA - It is used to determine one factor with atleast 2 levels and these levels are dependent here.
Factorial ANOVA - Used to determine 2 or more factors each of which with atleast 2 levels and these levels can be independent or depented

Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

ANOVA is based on the law of total variance, where the observed variance in a particular variable is partitioned into components attributable to different sources of variation. In its simplest form, ANOVA provides a statistical test of whether two or more population means are equal, and therefore generalizes the t -test beyond two means.

Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual
sum of squares (SSR) in a one-way ANOVA using Python?

1.Sum of Squares Total (SST) – The sum of squared differences between individual data points (yi) and the mean of the response variable (y).

SST = Σ(yi – y)2

2.Sum of Squares Error (SSE) – The sum of squared differences between predicted data points (ŷi) and observed data points (yi).

SSE = Σ(ŷi – yi)2

3.Sum of Squares Regression (SSR) – The sum of squared differences between predicted data points (ŷi) and the mean of the response variable(y).

SSR = Σ(ŷi – y)2


Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

Main effects deal with each factor separately. For a example if we have two factors, A and B. 
The main effect of Factor A is the difference between the mean for Specimen 1 and Specimen 2, averaged across the three levels of factor B. The main effect of Factor B is the difference in mean for levels 1, 2, and 3 averaged across the two species. 

The interaction is the simultaneous changes in the levels of both factors. If the changes in the level of Factor A result in different changes in the value of the response variable for the different levels of Factor B, we say that there is an interaction effect between the factors. 

Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02.
What can you conclude about the differences between the groups, and how would you interpret these
results?

Since this p-value is less than 0.05, we need to reject the null hypothesis.
This means that there is a statistically significant difference between the mean of the groups

Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential
consequences of using different methods to handle missing data?


One of the most effective ways of dealing with missing data is multiple imputation (MI), either using mean, median,mode depending on the case.
The potential consequences of using different methods to handle missing data depends on the outliers, categorical data.
If no outlier in data - Use mean
If outlier is there - use Median
If categorical data - Mode

Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide
an example of a situation where a post-hoc test might be necessary.

The different types of post hoc tests include Tukey’s HSD, Scheffe’s test, Bonferroni test, and Dunn’s test.

The Tukey post-hoc test should be used when you would like to make pairwise comparisons between group means when the sample sizes for each group are equal.

The Scheffe post-hoc test should be used when you would like to make all possible contrasts between group means. This test allows you to compare more than just two means at once, unlike the Tukey post-hoc test.

The Bonferroni post-hoc test should be used when you have a set of planned comparisons you would like to make beforehand. 

Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from
50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python
to determine if there are any significant differences between the mean weight loss of the three diets.
Report the F-statistic and p-value, and interpret the results.



In [11]:
import numpy as np
diet1 = np.random.randint(60, 85, 17)
diet2 = np.random.randint(70, 95, 17)
diet3 = np.random.randint(50, 80, 16)
print(diet1)
print(diet2)
print(diet3)

from scipy.stats import f_oneway
import scipy.stats as stat
#perform one-way ANOVA
f_oneway(diet1, diet2, diet3)



[75 84 72 82 66 60 65 81 64 80 74 63 63 68 69 75 73]
[88 91 80 80 70 78 89 92 88 77 75 86 93 70 78 70 91]
[54 61 67 63 60 51 50 70 56 78 78 51 74 60 74 71]


F_onewayResult(statistic=19.859437491014752, pvalue=5.605692070728685e-07)

Since Pvalue is not less than 0.05, we fail to reject the null hypothesis. There is no siginificant difference between 3 diets.

Q10. A company wants to know if there are any significant differences in the average time it takes to
complete a task using three different software programs: Program A, Program B, and Program C. They
randomly assign 30 employees to one of the programs and record the time it takes each employee to
complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or
interaction effects between the software programs and employee experience level (novice vs.
experienced). Report the F-statistics and p-values, and interpret the results.



In [18]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
import numpy as np
import pandas as pd

# Create a dataframe
df = pd.DataFrame({'Programs': np.repeat(['Program A', 'Program B', 'Program C'], 10),
                   'Experience': np.tile(np.repeat(['experienced', 'Novice', ], 5), 3),
                   'Time': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})
  
model = ols('Time ~ C(Programs) + C(Experience) + C(Programs):C(Experience)', data=df).fit()
sm.stats.anova_lm(model, typ=2)
 


sum_sq    df      F        PR(>F)
C(Programs)                 0.266667   2.0   0.25  7.808037e-01
C(Experience)               1.200000   1.0   2.25  1.466556e-01
C(Programs):C(Experience)  34.400000   2.0  32.25  1.582033e-07
Residual                   12.800000  24.0    NaN           NaN

Since the p-values for Programs and Experience are both larger than .05, this means that both factors have a statistically insignificant effect on Average time.

And since the p-value for the interaction effect (.000000158203) is less than .05, this tells us that there is a significant interaction effect between Programs variety and Experience level.

Q11. An educational researcher is interested in whether a new teaching method improves student test
scores. They randomly assign 100 students to either the control group (traditional teaching method) or the
experimental group (new teaching method) and administer a test at the end of the semester. Conduct a
two-sample t-test using Python to determine if there are any significant differences in test scores
between the two groups. If the results are significant, follow up with a post-hoc test to determine which
group(s) differ significantly from each other.

In [28]:
import scipy.stats as stats
 
# Creating data groups
traditional_teaching = np.random.randint(50,90,50)
new_teaching = np.random.randint(80,90,50)

 
print(np.var(traditional_teaching), np.var(new_teaching))
# Perform the two sample t-test with equal variances
stats.ttest_ind(a=data_group1, b=data_group2, equal_var=True)

125.59560000000002 9.515600000000001


Ttest_indResult(statistic=-0.6337397070250238, pvalue=0.5300471010405257)

The p-value of the test comes out to be equal to 0.530, which is greater than the significance level alpha (that is, 0.05). This implies that we can say that there is no different in level of teaching between traditional and new teaching methods. 

Q12. A researcher wants to know if there are any significant differences in the average daily sales of three
retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store
on those days. Conduct a repeated measures ANOVA using Python to determine if there are any
significant differences in sales between the three stores. If the results are s

In [25]:
import numpy as np
import pandas as pd
from statsmodels.stats.anova import AnovaRM
  

# Create the data
df = pd.DataFrame({'Store': np.repeat(['Store1','Store2','Store3'], 30),
                   'days': np.tile(np.arange(1, 31), 3),
                   'Sales': np.random.randint(30,45,90)})
 # Print the dataframe
print(dataframe)

# Conduct the repeated measures ANOVA
print(AnovaRM(data=df, depvar='Sales',
              subject='Store', within=['days']).fit())

    Store1  Store2  Store3
0       41      35      44
1       31      34      35
2       35      41      31
3       39      34      40
4       30      41      43
5       41      32      41
6       40      37      32
7       30      38      43
8       35      35      44
9       40      38      32
10      43      33      40
11      30      44      44
12      42      39      30
13      37      37      30
14      40      40      41
15      30      31      31
16      32      35      41
17      34      37      43
18      41      37      41
19      34      31      34
20      39      36      40
21      44      43      40
22      33      31      43
23      41      41      38
24      34      42      43
25      39      39      33
26      40      43      30
27      30      44      30
28      40      31      31
29      42      39      30
               Anova
     F Value  Num DF  Den DF Pr > F
-----------------------------------
days  1.1631 29.0000 58.0000 0.3062



In this example, the F test-statistic is 1.1631 and the corresponding p-value is 0.3062. Since this p-value is greater than 0.05, we fail to reject the null hypothesis and conclude that there is statistically no significant difference in mean sales between the three stores.